### PR TITLE
remove stacking operation from batched functions

### PR DESCRIPTION
**Pull request by lilithgrigoryan**
# What does this PR do ?

PR removed torch.stack calls from batched function yielding to faster decoding

**Collection**:
ASR

# Changelog 
### Refactoring
- **`batch_initialize_states()`** renamed to **`batch_stack_states`**
  - Reason: The function does not initialize any states, so the name was changed for clarity.
- **`batch_initialize_states()`** minimized torch stack calls
  - Reason: Torch stack internally copies the data.
- **`batch_select_state()`**
  - Reason: Function is hard to read, tried to make it clearer.
  
### Argument Removal
- **`batch_initialize_states()`**: Removed the `batch_states` argument as it was unused.
- **`batch_score_hypothesis()`**: Removed the `batch_states` argument as it was unused.

### Removing redundant calculations
- **`batch_score_hypothesis()`**: removed torch.stack calls since they are not needed. Returning python lists instead.
- **`batch_score_hypothesis()`**: removed `lm_tokens` return value as it was not used.

## Changes in `[nemo/collections/asr/modules/rnnt_abstract.py]`
- reflected comments and argument list changes in derived classes here

## Changes in `[nemo/collections/asr/parts/submodules/rnnt_beam_decoding.py]`
- replaced **`batch_select_state()`** calls with index access as now everything is a python list
- **`recombine_hypotheses()`** returning final hypotheses instead of initial ones.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/modules/rnnt.py [hunk 1]
```diff
@@ -317,20 +315,18 @@ def initialize_state(self, y: torch.Tensor) -> List[torch.Tensor]:
         ]
         return state
 
-    def batch_initialize_states(self, batch_states: List[torch.Tensor], decoder_states: List[List[torch.Tensor]]):
+    def batch_stack_states(self, decoder_states: List[List[torch.Tensor]]):
```
Comment (titu1994) [inline nemo/collections/asr/modules/rnnt.py:318]: Do we need to change the name? Many subclasses depend on this to override

**Review by titu1994 — COMMENTED** [review]
I'm a bit hesitant to change names of functions, but as long as the code passes the RNNT beam search tests it might be ok. @hainan-xv and @andrusenkoau for final review since they have the most recent experience with beam search for RNNT

File: nemo/collections/asr/modules/rnnt.py (hunk 1 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/modules/rnnt.py]: Thanks. Restored the original name

**Comment by lilithgrigoryan** [comment]
jenkins

**Review by andrusenkoau — APPROVED** [review]
LGTM, thanks for fixing the errors!

File: nemo/collections/asr/modules/rnnt.py [hunk 2]
```diff
@@ -1057,15 +1015,11 @@ def batch_select_state(self, batch_states: List[torch.Tensor], idx: int) -> List
             (tuple): decoder states for given id
                 ([L x (1, H)], [L x (1, H)])
         """
+        # print("###", len(batch_states), batch_states[0].shape, self.pred_rnn_layers)
```
Comment (hainan-xv) [inline nemo/collections/asr/modules/rnnt.py:1018]: this should be removed

**Review by hainan-xv — COMMENTED** [review]
Looks mostly good to me! I left some small comments.

File: nemo/collections/asr/modules/rnnt_abstract.py [hunk 3]
```diff
@@ -226,41 +226,37 @@ def score_hypothesis(
...
-           batch_states (list): batch of decoder states
-              ([L x (B, H)], [L x (B, H)])
-
-           decoder_states (list of list): list of decoder states
-               [B x ([L x (1, H)], [L x (1, H)])]
+        Args:
+            decoder_states (list of list of list of torch.Tensor): list of decoder states
+                [B, L, 1, H]
+                    - B: Batch size.
+                    - L: Number of layers in prediction RNN (e.g., for LSTM, this is 2: hidden and cell states).
```
[the remark this answers is not in the document]
Comment (lilithgrigoryan) [reply nemo/collections/asr/modules/rnnt_abstract.py:253]: Thanks! Corrected

File: nemo/collections/asr/modules/rnnt.py (hunk 2 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/modules/rnnt.py:1018]: Thanks!

**Review by titu1994 — APPROVED** [review]
Looks ok, thanks for the refactor ! Can you add some inline docs to the new steps for future reference?
